File: src/web/types/task.ts
export type TaskType = 'code_execution' | 'file_processing' | 'network_access' | 'data_analysis' | 'text_processing' | 'custom';

export type TaskStatus = 'pending' | 'running' | 'completed' | 'failed' | 'paused' | 'stopped';

export type TaskPriority = 'low' | 'medium' | 'high' | 'critical';

export interface TaskStep {
  id: string;
  taskId: string;
  name: string;
  description?: string;
  status: TaskStatus;
  order: number;
  result?: any;
  error?: string;
  created_at: string;
  updated_at: string;
}

export interface TaskResult {
  success: boolean;
  data: any;
  error?: string;
  metrics?: {
    duration: number;
    resourceUsage: {
      cpu: number;
      memory: number;
    };
  };
}

export interface Task {
  id: string;
  name: string;
  type: TaskType;
  status: TaskStatus;
  priority: TaskPriority;
  description?: string;
  config?: Record<string, any>;
  steps: TaskStep[];
  result?: TaskResult;
  agent_id?: string;
  created_at: string;
  updated_at: string;
}

export interface TaskFilter {
  type?: TaskType[];
  status?: TaskStatus[];
  priority?: TaskPriority[];
  agentId?: string;
  startDate?: string;
  endDate?: string;
  search?: string;
}

export interface TaskStats {
  total: number;
  completed: number;
  failed: number;
  running: number;
  pending: number;
  byType: Record<TaskType, number>;
  byStatus: Record<TaskStatus, number>;
  byPriority: Record<TaskPriority, number>;
  averageCompletionTime: number;
  successRate: number;
}

export interface TaskState {
  tasks: Record<string, Task>;
  selectedTask: Task | null;
  filters: TaskFilter;
  stats: TaskStats;
  loading: boolean;
  error: string | null;
}

export type TaskAction =
  | { type: 'SET_TASKS'; payload: Record<string, Task> }
  | { type: 'SELECT_TASK'; payload: Task }
  | { type: 'UPDATE_TASK'; payload: Task }
  | { type: 'DELETE_TASK'; payload: string }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string }
  | { type: 'ADD_TASK'; payload: Task }
  | { type: 'CLEAR_ERROR' }; 

File: src/models/task.py
from sqlalchemy import Column, String, Integer, Text, JSON
import enum
from .base import BaseModel

class TaskType(enum.Enum):
    CODE_EXECUTION = "code_execution"
    DATA_ANALYSIS = "data_analysis"
    NETWORK_REQUEST = "network_request"
    SYSTEM_COMMAND = "system_command"

class TaskStatus(enum.Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"

class TaskPriority(enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"

class Task(BaseModel):
    """Task model for storing task information."""
    __tablename__ = "tasks"

    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    type = Column(String(50), nullable=False)
    status = Column(String(20), nullable=False, default=TaskStatus.PENDING.value)
    priority = Column(String(20), nullable=False, default=TaskPriority.MEDIUM.value)
    task_metadata = Column(JSON, nullable=True)
    result = Column(JSON, nullable=True)
    error = Column(Text, nullable=True)
    progress = Column(Integer, nullable=False, default=0)

    def to_dict(self):
        """Convert task instance to dictionary."""
        result = super().to_dict()
        return result

    @classmethod
    def from_dict(cls, data: dict):
        """Create task instance from dictionary."""
        if 'type' in data and isinstance(data['type'], TaskType):
            data['type'] = data['type'].value
        if 'status' in data and isinstance(data['status'], TaskStatus):
            data['status'] = data['status'].value
        if 'priority' in data and isinstance(data['priority'], TaskPriority):
            data['priority'] = data['priority'].value
        return super().from_dict(data) 